### ISSUE TITLE
controller lease watcher watches from outdated index

### ISSUE THREAD

**Issue by liggitt** [openshift/origin#11617]
the controller lease watcher watches from the index of the lease etcd node:

```go
    nextIndex = eventIndexFor(latest)
    if latest.Node.TTL > 0 {
        ttl = uint64(latest.Node.TTL)
    }

    if latest.Node.Value != e.value {
        glog.V(4).Infof("Lease %s owned by %s at %d ttl %d seconds, waiting for expiration", e.key, latest.Node.Value, nextIndex-1, ttl)
        // waits until the lease expires or changes to us.
        // TODO: it's possible we were given the lease during the watch, but we just expect to go
        //   through this loop again and let this condition check
        if _, err := e.waitForExpiration(false, nextIndex, nil); err != nil {
            return false, 0, 0, fmt.Errorf("unable to wait for lease expiration %s: %v", e.key, err)
        }
        return false, 0, 0, nil
    }
```

if that index is outside the watch window, the watch will immediately fail. the controller retries, but keeps retrying from the index of the node (rather than the current cluster index)

this means the lease watching devolves into a 1 second poll, rather than a watch.

**Comment by pweil-** [comment]
Currently a 3.4 blocker, is this going to make Friday? 

@mfojtik FYI

**Comment by mfojtik** [comment]
@pweil- @kargakis will have look

**Comment by liggitt** [comment]
Not new in 3.4, not convinced it's a blocker. The fix *should* be small, but we might want more soak time.

**Comment by liggitt** [comment]
since this isn't a regression, and I don't have evidence that the lease poll causes problems, I'd recommend changing to p2 and trying to fix it at the beginning of a release

to test, we'd need to make sure we had a write rate high enough (or a lease time long enough) that the watch window is exceeded before the lease renewal

**Comment by php-coder** [comment]
@liggitt is it planning for 3.5 or it's too late?

**Comment by liggitt** [comment]
probably too late for 1.5, but can you work on prepping the fix so we can drop it at the beginning of 1.6 for soak time? Can you also work on reproducing the issue (might be hard to do in an automated test, but should be easy to do manually with a long lease and high-rate etcd write) so we can verify the fix works properly and lease failover still works?

**Comment by liggitt** [comment]
controllers no longer hit etcd directly for leases

/close
